text/0000-custom-model-classes.md

```diff
@@ -0,0 +1,225 @@
+
+In order for the record class to learn about changes to the underlying data, it will be passed a `NotificationManager` as a constructor argument, which will allow it to
+subscribe and unsubscribe to notifications of underlying changes to the data. Once `RecordData` calls one of the notification methods, the notification manager will call 
+any registered callback for the given identifier, and pass in the type of the notification, allowing the record the opportunity to repull the data if needed. There are no guarantees around the timing of the notification callback being called after `RecordData` informs the store of changes. We expect that in a modern Ember app with tracked properties, this wouldn't be the common path for tracking changes. 
+
+```ts
+interface UnsubscribeToken {
+  unsubscribe(): void;
+}
+
+interface NotificationCallback {
+  (identifier: Identifier, notificationType: 'attributes' | 'relationships' | 'errors'  | 'meta'  | 'unload'): void;
+}
+    
+interface NotificationManager {
+  subscribe(identifier, NotificationCallback): UnsubscribeToken
+ }
+```
+
+## Exposing schema information
+
+We currently keep schema information on `DS.Model` class. In order to allow for custom Model implementations we need to allow lookup of schema info from the store. We already have specified a schema api that RecordData consumes: `attributesDefinitionFor` and `relationshipDefinitionFor`. We would define these methods publicly on the store for addons to use. **These methods are not ergonomic on purpose.** They match the current Record Data apis and are designed as a stepping stone on the way of having a better, user facing schema APIs. The schema info is currently primarily geared towards being used by the internal relationship handling layer, the serializer/adapter layers and the DebugAdapter. Schema methods support both static and dynamic schema computation. For static schemas, the method can always respond with a schema definition based on the type passed, and for dynamic changing schemas, it can look up the underlying data by the identifer which is passed in. We would also add a method called `isModelType`, which would return `true` if ED knew that a given string is a model type and `false` otherwise.
+  
+
+```ts
+export interface RelationshipDefinition {
+  kind: 'hasMany'| 'belongsTo';
+  type: string;
+  options: { [key: string]: any } ;
+  name: string;
+}
+    
+interface RelationshipsDefinition {
+  [key: string]: RelationshipDefinition
+}
+    
+interface AttributeDefinition {
+  name: string;
+  options: { [key: string]: any };
+  type: string;
+}
+
+interface AttributesDefinition {
+  [key: string]: AttributeDefinition
+}
+    
+class Store {
+  // Following the existing RD implementation 
+  attributesDefitionFor(identifier: RecordIdentifier | type: string): AttributesDefiniton
+  
+  // Following the existing RD implementation
+  relationshipsDefinitionFor(identifier: RecordIdentifier | type: string): RelationshipsDefinition
```

Comment:
What does it actually look like for a userland model class to provide schema information this way?  Are users expected to overwrite these methods in their own `store`?

The current implementation here has the store knowing something intimate about `DS.Model`. (in particular that it can read `attributes` and `relationshipsObject` from the model class).  Should we split the store into the "core" store and "default model userland" store?

Reply:
You would overwrite the methods. If you were doing that, you wouldn't be reading off of the DS.Model presumably. Not quite sure I understand the store split proposal, let's discuss it at the meeting

Reply:
I would like to split the store.

Reply:
(That discussion was rapidly approaching and having it now seems a good time)

Reply:
The suggestion is for there to be a `store.ts` file that does not know anything about `DS.Model`, in part as a way to dogfood that `DS.Model` is actually something that can be swapped out.